src/org/rascalmpl/values/RascalFunctionValueFactory.java

```diff
@@ -567,10 +574,24 @@ protected IValue parse(String methodName, ISet filters, ISourceLocation input, I
             }
         }
 
-        private ITree parseObject(String methodName, ISourceLocation location, char[] input,  boolean allowAmbiguity, boolean hasSideEffects,  ISet filters) {
+        private ITree parseObject(String methodName, ISourceLocation location, char[] input,  boolean allowAmbiguity, boolean allowRecovery, boolean hasSideEffects,  ISet filters) {
             IActionExecutor<ITree> exec = filters.isEmpty() ?  new NoActionExecutor() : new RascalFunctionActionExecutor(filters, !hasSideEffects);
+            IGTD<IConstructor, ITree, ISourceLocation> parserInstance = getParser();
+            IRecoverer<IConstructor> recoverer = null;
+            IDebugListener<IConstructor> debugListener = null;
+            URI uri = location.getURI();
+            if (allowRecovery) {
+                recoverer = new ToTokenRecoverer(uri, parserInstance, new StackNodeIdDispenser(parserInstance));
+            }
+            ITree parseForest = (ITree) parserInstance.parse(methodName, uri, input, exec, new DefaultNodeFlattener<>(), new UPTRNodeFactory(allowRecovery || allowAmbiguity), recoverer, debugListener);
+
+            if (!allowAmbiguity && allowRecovery && filters.isEmpty()) {
+                // Filter error-induced ambiguities
+                RascalValueFactory valueFactory = (RascalValueFactory) ValueFactoryFactory.getValueFactory();
+                parseForest = (ITree) new ErrorRecovery(valueFactory).disambiguateErrors(parseForest, valueFactory.bool(false));
```

Comment:
I think this disambiguation step is smart in most cases, but it should be optional and on-by-default. When we do repair or autocomplete of parse errors, then some of the filtered ambiguities could have been insightful for the user. A comparable algorithm would _rank_ the alternatives, for example by the amount of skipped tokens; there are many different options which should be left to the language designer.

Reply:
What happens if we skip this step here and leave it to the language designer in their parse function, or their services?

Reply:
If you set `allowAmbiguity=true` a language designer already gets the opportunity to filter the trees themself. So it's already optional and on-by-default.

Reply:
I feel the allowAmbiguity flag is meant to indicate something different. This option now conflates error recovery for a user with developer ambiguity in a grammar. 

It's important for usability (performance) to still be able to exit tree construction quickly on  unexpected ambiguity, _and_ have predictable error recovery that represents all currently valid prefixes. 

I don't know what the right solution is yet; but I think that allowAmbiguity=false should be ignored after error recovery

Reply:
To make this less subtle-sounding:
1. A normal ambiguity is almost _never intended_, confuses the downstream tools as well as the user, and possibly leads to high polinomial tree constructing time (very slow). Typically we disallow ambiguity after grammar deployment because of this, while we need ambiguity to debug grammars at development time.
2. An error ambiguity is _almost always inherent_ to the recovery process. The clusters represent the different viable prefixes at the moment the parser discovered it was stuck. The use needs all of them for useful features like auto-complete and quickfix. Disallowing those after deployment will hamper the usability of error recovery severely. 

In other words we do expect error ambiguïty but we don't expect grammatical ambiguity. So it can't be the same option. 

It's good to know, as a kind of illustration, that error recovery for a single typo will produce ambiguous clusters of predictions on different levels, for different nonterminals in the tree. At the same time also on the same line, and same nonterminals, different prefixes could be active. As a result one recovered file {w,c,sh}ould receive quickfix proposals on different line/column positions. But if we already filtered heuristically because "ambiguity is not allowed", many of the valid and natural options will have been removed.

Reply:
We could pass in an external error filtering function similar to the actions, of that is required for speed, but otherwise I think it would be best to not filter at this stage and leave the ranking to quickfix and autocomplete.

Reply:
These are compelling arguments. So we should remove the automatic error tree disambiguation altogether and leave it to the developer. A developer can always specify a filter to do the disambiguation.

Currently the disambiguation function takes two arguments (the tree and a boolean specifying if "normal" ambiguities are allowed). Maybe we should provide two separate functions that only take a tree argument so the programmar can just use one of them directly as a parse filter?

Reply:
Yes. We can have examples of reusable functions and filters in Recover.rsc

Reply:
Thanks for the discussion!

1. It's exactly the normal users who have to be protected against heuristic filters. Their semantics is unpredictable while if we built it in it becomes a contract for their downstream tooling. The shortest error is truly nothing more than a blunt heuristic which misfires more than it doesn't. Especially if we want to provide grammar-derived repair for those simpler users the shortest error makes little sense, but I'm pretty sure it almost never is the "right" one.

We've been here before in the 2000's with "powerful" disambiguation filters. Removing them because of their semantic inaccuracy was a social-economic wasteland. People were not happy their trees "suddenly" became ambiguous, while they were fully ignorant of what they couldn't be aware of: the trees they actually should care about. Hiding ambiguity is a mine field that I want to stay clear of. 

2. That's not true. The forest represents all the viable prefixes at the time of getting stuck. None of them are more likely to be the intended future representation, after repair, in general. It's worse even. Many of the predictions may have already failed and garbage collected, making the forest also incomplete in that sense. More incompleteness will make this only worse. Currently we have a declarative and complete contract for the output of error recovery; with any auto-filter the contract becomes ill-defined from a language semantics perspective.

3. Error recovery and repair or diagnostics algorithmes do not belong in the `parser` contribution, ever. The requirements on the downstream processor dictate how to deal with errors or multiple ambiguous errors. A repair Algorithms takes each prefix to complete, a type checker may want to skip entire expressions and statements that contain errors at all, a diagnostics algorithm may fuzz around the error positions and reparse, etc. Different filters for different backends. Filtering errors is not a syntax aspect, it's a semantics aspect.

In one LSP server, different services will deal differently with the errors and their clusters. The first priority should be to make sure that algorithms written by beginners fail either gracefully or stay accidentally robust in the presence of errors. Features like `visit` and dynamic dispatch already deal gracefully with ambiguity and errors (they simply don't match). To make Algorithms robust without thinking at ask all the users must do is add a default case that does nothing.

The dirty edges w still have the brush up are Rascal features that look into trees without pattern matching. See the pr on Field projection. that solution still has edge cases that I'd like to see buttoned up.